File: js/passwordGenerator.js
const LOWERCASE = 'abcdefghijklmnopqrstuvwxyz'
const UPPERCASE = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
const DIGITS = '0123456789'
const SPECIAL = '!@#$%^&*'

const nextRandom = (seed) => (16807 * seed) % 2147483647

const generatePassword = (length, seed, options = {}) => {
  const { useUppercase = true, useDigits = true, useSpecial = false } = options

  let alphabet = LOWERCASE
  if (useUppercase) alphabet += UPPERCASE
  if (useDigits) alphabet += DIGITS
  if (useSpecial) alphabet += SPECIAL

  // Math.abs перед %, т.к. в JS % отрицательного числа = отрицательный результат
  let current = Math.abs(seed) % 2147483647
  if (current === 0) current = 1

  let result = ''
  for (let i = 0; i < length; i += 1) {
    current = nextRandom(current)
    result += alphabet[current % alphabet.length]
  }
  return result
}

const hasLowercase = (password) => {
  for (const char of password) {
    if (char >= 'a' && char <= 'z') return true
  }
  return false
}

const hasUppercase = (password) => {
  for (const char of password) {
    if (char >= 'A' && char <= 'Z') return true
  }
  return false
}

const hasDigit = (password) => {
  for (const char of password) {
    if (char >= '0' && char <= '9') return true
  }
  return false
}

const hasSpecial = (password) => {
  for (const char of password) {
    if (SPECIAL.includes(char)) return true
  }
  return false
}

const isLongEnough = (password, minLength = 8) => password.length >= minLength

const strengthScore = (password) => {
  let score = 0
  if (isLongEnough(password)) score += 1
  if (hasLowercase(password)) score += 1
  if (hasUppercase(password)) score += 1
  if (hasDigit(password)) score += 1
  if (hasSpecial(password)) score += 1
  return score
}

const checkPassword = (password) => {
  const score = strengthScore(password)
  let verdict
  if (score <= 2) {
    verdict = 'Слабый'
  } else if (score === 3) {
    verdict = 'Средний'
  } else if (score === 4) {
    verdict = 'Надёжный'
  } else {
    verdict = 'Очень надёжный'
  }
  return `${verdict} пароль (оценка ${score} из 5)`
}

export { generatePassword, checkPassword }

File: go/password_generator.go
package code

import "fmt"

const (
	lowercase = "abcdefghijklmnopqrstuvwxyz"
	uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	digits    = "0123456789"
	special   = "!@#$%^&*"
)

// NextRandom возвращает следующее псевдослучайное число по формуле LCG.
func NextRandom(seed int) int {
	return (16807 * seed) % 2147483647
}

// GeneratePassword генерирует пароль заданной длины по числу-ключу.
func GeneratePassword(length, seed int, useUppercase, useDigits, useSpecial bool) string {
	alphabet := lowercase
	if useUppercase {
		alphabet += uppercase
	}
	if useDigits {
		alphabet += digits
	}
	if useSpecial {
		alphabet += special
	}

	// Go: % от отрицательного числа = отрицательный результат → берём модуль вручную
	if seed < 0 {
		seed = -seed
	}
	current := seed % 2147483647
	if current == 0 {
		current = 1
	}

	result := ""
	for i := 0; i < length; i++ {
		current = NextRandom(current)
		result += string(alphabet[current%len(alphabet)])
	}
	return result
}

func hasLowercase(password string) bool {
	for _, c := range password {
		if c >= 'a' && c <= 'z' {
			return true
		}
	}
	return false
}

func hasUppercase(password string) bool {
	for _, c := range password {
		if c >= 'A' && c <= 'Z' {
			return true
		}
	}
	return false
}

func hasDigit(password string) bool {
	for _, c := range password {
		if c >= '0' && c <= '9' {
			return true
		}
	}
	return false
}

func hasSpecial(password string) bool {
	for _, c := range password {
		for _, s := range special {
			if c == s {
				return true
			}
		}
	}
	return false
}

func isLongEnough(password string, minLength int) bool {
	return len(password) >= minLength
}

func strengthScore(password string) int {
	score := 0
	if isLongEnough(password, 8) {
		score++
	}
	if hasLowercase(password) {
		score++
	}
	if hasUppercase(password) {
		score++
	}
	if hasDigit(password) {
		score++
	}
	if hasSpecial(password) {
		score++
	}
	return score
}

// CheckPassword оценивает надёжность пароля и возвращает вердикт.
func CheckPassword(password string) string {
	score := strengthScore(password)
	var verdict string
	switch {
	case score <= 2:
		verdict = "Слабый"
	case score == 3:
		verdict = "Средний"
	case score == 4:
		verdict = "Надёжный"
	default:
		verdict = "Очень надёжный"
	}
	return fmt.Sprintf("%s пароль (оценка %d из 5)", verdict, score)
}
